gasmodel/Pact/Core/GasModel/Utils.hs

```diff
@@ -322,21 +324,80 @@ runNativeBenchmark
   -> C.Benchmark
 runNativeBenchmark = runNativeBenchmark' pure pure
 
-runNativeBenchmarkPrepared
-  :: [(Text, PactValue)]
-  -> PactDb CoreBuiltin ()
-  -> String
-  -> Text
-  -> C.Benchmark
-runNativeBenchmarkPrepared envVars = runNativeBenchmark' pure stMod
+withLoaded :: [(Text, PactValue)] -> (BenchEvalState -> BenchEvalState)
+withLoaded envVars = esLoaded .~ synthLoaded
   where
   synthLoaded = Loaded
     { _loModules = mempty
     , _loToplevel = M.fromList [ (n, (mkGasModelFqn n, DKDefConst)) | n <- fst <$> envVars ]
     , _loNamespace = Nothing
     , _loAllLoaded = M.fromList [ (mkGasModelFqn n, DConst $ DefConst n Nothing (EvaledConst v) ()) | (n, v) <- envVars ]
     }
-  stMod = pure . (esLoaded .~ synthLoaded)
+
+runNativeBenchmarkPrepared
+  :: [(Text, PactValue)]
+  -> PactDb CoreBuiltin ()
+  -> String
+  -> Text
+  -> C.Benchmark
+runNativeBenchmarkPrepared envVars = runNativeBenchmark' pure (pure . withLoaded envVars)
+
+type EnvMod = Endo BenchEvalEnv
+
+msgBody :: Map Field PactValue -> EnvMod
+msgBody body = Endo $ eeMsgBody .~ PObject body
+
+msgSigsNoCap :: [PublicKeyText] -> EnvMod
+msgSigsNoCap pkts = Endo $ eeMsgSigs .~ M.fromList ((, mempty) <$> pkts)
+
+runNativeBenchmarkPreparedEnvMod
+  :: EnvMod
+  -> [(Text, PactValue)]
+  -> PactDb CoreBuiltin ()
+  -> String
+  -> Text
+  -> C.Benchmark
+runNativeBenchmarkPreparedEnvMod (Endo envMod) envVars = runNativeBenchmark' (pure . envMod) (pure . withLoaded envVars)
+
+type StMod = Endo BenchEvalState
+
+stCaps :: [CapToken QualifiedName PactValue] -> StMod
+stCaps capToks = Endo $ esCaps.csSlots .~ ((`CapSlot` []) <$> capToks)
+
+stManaged :: [ManagedCap QualifiedName PactValue] -> StMod
+stManaged manageds = Endo $ esCaps.csManaged .~ S.fromList manageds
+
+stAddDef :: Text -> Def Name Type CoreBuiltin () -> StMod
+stAddDef name dfn = Endo $ (esLoaded.loToplevel %~ M.insert name (fqn, dkind))
+                         . (esLoaded.loAllLoaded %~ M.insert fqn dfn)
+  where
+  fqn = mkGasModelFqn name
+  dkind = case dfn of
+    Dfun{} -> DKDefun
+    DConst{} -> DKDefConst
+    DCap{} -> DKDefCap
+    DSchema{} -> error "stAddDef: todo support schemas"
```

Comment:
leftover todo

Reply:
Not yet! `DKDefSchema` requires a `Schema` (thus differing from the rest of `DefKind` ctors), and while I think it could be constructed from the definition itself, I didn't want to spend much time on this right now, since I don't have defschemas in my benchmarks so far. But shall we ever encounter those, we'll get an error message telling what to do!

If you feel like it's still worth addressing right now instead of evaluating lazily, lmk and I'll take a look.